File: stubs/Strand.php
<?php
/**
 * php-asio/stubs/Strand.php
 *
 * This file is a stub of php-asio, which is implemented in extension asio.so.
 * For source code of asio.so, see src/ directory.
 *
 * @author CismonX<admin@cismon.net>
 */

namespace Asio;

/**
 * Class Strand
 *
 * Provides serialised handler execution.
 *
 * @package Asio
 */
final class Strand
{
    /**
     * This class can only be instantiated using `Service::addStrand()`.
     */
    private function __construct() {}

    /**
     * Request the strand to invoke the given handler.
     *
     * @param callable $callback
     * @param mixed $argument
     * @return void
     */
    function dispatch(callable $callback, $argument = null) {}

    /**
     * Request the strand to invoke the given handler and return immediately.
     *
     * @param callable $callback
     * @param mixed $argument
     * @return void
     */
    function post(callable $callback, $argument = null) {}

    /**
     * Determine whether the strand is running in the current thread.
     *
     * @return bool
     */
    function runningInThisThread() {}

    /**
     * Create a new handler that automatically dispatches the wrapped handler on the strand.
     *
     * @param callable $callback[optional]
     * @return WrappedHandler
     */
    function wrap(callable $callback) {}
}

File: src/strand.cpp
/**
 * php-asio/strand.cpp
 *
 * @author CismonX<admin@cismon.net>
 */

#include "strand.hpp"

#ifdef ENABLE_STRAND
namespace Asio
{
    Strand::Strand(boost::asio::io_service& io_service) : strand_(io_service)
    {
        wrapper_ = new Php::Object("Asio\\Strand", this);
    }

    void Strand::dispatch(Php::Parameters& params)
    {
        auto callback = params[0];
        auto argument = params.size() == 1 ? Php::Value() : params[1];
        strand_.dispatch([callback, argument]()
        {
            CORO_REGISTER(callback(argument));
        });
    }

    void Strand::post(Php::Parameters& params)
    {
        auto callback = params[0];
        auto argument = params.size() == 1 ? Php::Value() : params[1];
        strand_.post([callback, argument]()
        {
            CORO_REGISTER(callback(argument));
        });
    }

    Php::Value Strand::running_in_this_thread() const
    {
        return strand_.running_in_this_thread();
    }

    Php::Value Strand::wrap(Php::Parameters& params)
    {
        return Php::Object("Asio\\WrappedHandler",
            new WrappedHandler(strand_, params.size() ? params[0] : Php::Value()));
    }

    void Strand::destroy()
    {
        if (wrapper_) {
            delete wrapper_;
            wrapper_ = nullptr;
        }
    }
}
#endif // ENABLE_STRAND
